packages/hydrogen/src/pagination/Pagination.ts

```diff
@@ -0,0 +1,263 @@
+  state: {
+    nodes: Array<NodesType>;
+    pageInfo: {
+      endCursor: Maybe<string> | undefined;
+      startCursor: Maybe<string> | undefined;
+      hasPreviousPage: boolean;
+    };
+  };
+}
+
+type PaginationProps<NodesType> = {
+  /** The response from `storefront.query` for a paginated request. Make sure the query is passed pagination variables and that the query has `pageInfo` with `hasPreviousPage`, `hasNextpage`, `startCursor`, and `endCursor` defined. */
+  connection: Connection<NodesType>;
+  /** A render prop that includes pagination data and helpers. */
+  children: PaginationRenderProp<NodesType>;
+};
+
+type PaginationRenderProp<NodesType> = (
+  props: PaginationInfo<NodesType>,
+) => JSX.Element | null;
+
+export function Pagination<NodesType>({
+  connection,
+  children = () => null,
+}: PaginationProps<NodesType>) {
+  const transition = useNavigation();
+  const isLoading = transition.state === 'loading';
+  const {
+    endCursor,
+    hasNextPage,
+    hasPreviousPage,
+    nextPageUrl,
+    nodes,
+    previousPageUrl,
+    startCursor,
+  } = usePagination<NodesType>(connection);
+
+  const state = useMemo(
+    () => ({
+      pageInfo: {
+        endCursor,
+        hasPreviousPage,
+        startCursor,
+      },
+      nodes,
+    }),
+    [endCursor, hasPreviousPage, startCursor, nodes],
+  );
+
+  const NextLink = useMemo(
+    () =>
+      function NextLink(props: Omit<LinkProps, 'to'>) {
+        return hasNextPage
+          ? createElement(Link, {
+              preventScrollReset: true,
+              ...props,
+              to: nextPageUrl,
+              state,
+            })
+          : null;
+      },
+    [hasNextPage, nextPageUrl],
+  );
+
+  const PreviousLink = useMemo(
+    () =>
+      function PrevLink(props: Omit<LinkProps, 'to'>) {
+        return hasPreviousPage
+          ? createElement(Link, {
+              preventScrollReset: true,
+              ...props,
+              to: previousPageUrl,
+              state,
+            })
+          : null;
+      },
+    [hasPreviousPage, previousPageUrl],
+  );
+
+  return children({
+    state,
+    hasNextPage,
+    hasPreviousPage,
+    isLoading,
+    nextPageUrl,
+    nodes,
+    previousPageUrl,
+    NextLink,
+    PreviousLink,
+  });
+}
+
+/**
+ * Get cumulative pagination logic for a given connection
+ */
+export function usePagination<NodesType>(
+  connection: Connection<NodesType>,
+): Omit<
+  PaginationInfo<NodesType>,
+  'isLoading' | 'state' | 'NextLink' | 'PreviousLink'
+> & {
+  startCursor: Maybe<string> | undefined;
+  endCursor: Maybe<string> | undefined;
+} {
+  const [nodes, setNodes] = useState(connection.nodes);
+  const {state, search} = useLocation() as {
```

Comment:
can we do `useLocation<{type here}>()` instead?

Reply:
Sadly, nope, `useLocation` doesn't take a generic. I think the cast is being used to coerce `state` to be the expected value that we saved. But it's incorrect because that state isn't guaranteed to be there. So it probably should instead have optional properties, and `PaginationState` should also have optional properties (forcing null checks):

```ts
  const {state, search} = useLocation() as {
    state?: PaginationState<NodesType>;
    search?: string;
  };
```